crates/dslab-iaas/src/experiments.rs

```diff
@@ -0,0 +1,175 @@
+    let mut logger: Box<dyn Logger> = Box::new(StdoutLogger::with_level(log_level.unwrap_or(Level::Info)));
+    if log_file.is_some() {
+        logger = Box::new(FileLogger::with_level(log_level.unwrap_or(Level::Info)));
+    }
+
+    let sim = Simulation::new(123);
+    let cloud_sim = rc!(refcell!(CloudSimulation::with_logger(
+        sim,
+        current_config.clone(),
+        logger
+    )));
+    callbacks.on_simulation_start(cloud_sim.clone());
+
+    while cloud_sim.borrow_mut().current_time() < current_config.simulation_length {
+        cloud_sim.borrow_mut().steps(1);
+        let proceed = callbacks.on_step(cloud_sim.clone());
+        if !proceed {
+            break;
+        }
+    }
+
+    let result = callbacks.on_simulation_finish(cloud_sim.clone());
+    if log_file.is_some() {
+        let save_result = cloud_sim.borrow_mut().save_trace_to_file(&log_file.clone().unwrap());
+        match save_result {
+            Ok(_) => println!("Log saved succesfully to file: {}", log_file.unwrap()),
+            Err(e) => println!("Error while saving log: {e:?}"),
+        }
+    }
+
+    result
+}
+
+pub struct Experiment {
+    pub callbacks: Box<dyn SimulationCallbacks>,
+    pub config: ExperimentConfig,
+    pub threads: usize,
+    pub log_file_prefix: Option<String>,
+    pub log_level: Option<Level>,
+}
+
+impl Experiment {
+    pub fn new(
+        callbacks: Box<dyn SimulationCallbacks>,
+        config: ExperimentConfig,
+        threads: usize,
+        log_file_prefix: Option<String>,
+        log_level: Option<Level>,
+    ) -> Self
+    where
+        Self: Sized,
+    {
+        Self {
+            callbacks,
+            config,
+            threads,
+            log_file_prefix,
+            log_level,
+        }
+    }
+
+    pub fn run(&mut self) {
+        let pool = ThreadPool::new(self.threads);
+        let test_cases = Arc::new(Mutex::new(Vec::new()));
+        let mut results: Vec<IndexMap<String, Dict>> = Vec::new();
+        let mut count = 1;
+
+        while let Some(current_config) = self.config.get() {
+            let config_info = format!("{0:?}", self.config);
+            let callbacks = self.callbacks.clone();
+            let log_level = self.log_level;
+
+            let mut log_file = self.log_file_prefix.clone();
+            if log_file.is_some() {
+                log_file = Some(format!("{}{}.csv", log_file.unwrap(), count));
+            }
+
+            //let config_fmt = current_config.clone();
+            let test_cases = test_cases.clone();
+            pool.execute(move || {
+                println!();
+                println!();
+                println!("==== New test case ====");
+                println!("{}", config_info);
+                println!("=======================");
+                println!();
+                println!();
+
+                test_cases.lock().unwrap().push(process_test_case(current_config.clone(), &mut callbacks.clone(), log_file, log_level));
```

Comment:
Выполнение `process_test_case` происходит под мьютексом (держим его все время пока не появится результат), то есть потоки работают по очереди. Надо вынести вычисление результата:

```suggestion
                let result = process_test_case(current_config.clone(), &mut callbacks.clone(), log_file, log_level);
                test_cases.lock().unwrap().push(result);
```